### PR TITLE
Added clamped and asynchronous loop handling for multiple animations

### PR DESCRIPTION
**Pull request by likangning93**
Fixes https://github.com/AnalyticalGraphicsInc/cesium/issues/5694

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Signed CLA is on file.

@likangning93, thanks for the pull request! Maintainers, we have a signed CLA from @likangning93, so you can review this at any time.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

File: Source/Core/CatmullRomSpline.js [hunk 1]
```diff
@@ -267,6 +267,24 @@ define([
      */
     CatmullRomSpline.prototype.findTimeInterval = Spline.prototype.findTimeInterval;
 
+    /**
+     * Wraps the given time to the period covered by the spline.
+     * @function
+     *
+     * @param {Number} time The time.
+     * @return {Number} the time, wrapped around to the updated animation.
```
Comment (lilleyse) [inline Source/Core/CatmullRomSpline.js]: For all these change to `The time`

File: Source/Core/Spline.js [hunk 2]
```diff
@@ -117,5 +119,55 @@ define([
...
+    /**
+     * Wraps the given time to the period covered by the spline.
+     * @function
+     *
+     * @param {Number} time The time.
+     * @return {Number} the time, wrapped around the animation period.
+     */
+    Spline.prototype.wrapTime = function(time) {
+        //>>includeStart('debug', pragmas.debug);
+        if (!defined(time)) {
```
Comment (lilleyse) [inline Source/Core/Spline.js]: Use `Check.typeOf.number` for these.

File: Source/Core/Spline.js [hunk 3]
```diff
@@ -117,5 +119,55 @@ define([
...
+            throw new DeveloperError('time is required.');
+        }
+        //>>includeEnd('debug');
+
+        // wrap
```
Comment (lilleyse) [inline Source/Core/Spline.js]: This and `// clamp` aren't needed.

File: Source/DataSources/ModelGraphics.js [hunk 4]
```diff
@@ -179,6 +181,14 @@ define([
          */
         runAnimations : createPropertyDescriptor('runAnimations'),
 
+        /**
+         * Gets or sets the boolean Property specifying if glTF animations should hold the last pose when no frames are specified.
```
Comment (lilleyse) [inline Source/DataSources/ModelGraphics.js]: For consistency make sure that these descriptions are the same. This and above are slightly different.

File: Source/Scene/Model.js [hunk 5]
```diff
@@ -2481,14 +2483,22 @@ define([
     }
 
     function getChannelEvaluator(model, runtimeNode, targetPath, spline) {
+        if (model._clampAnimations) {
+            return function(localAnimationTime) {
+                if (defined(spline)) {
+                    runtimeNode[targetPath] = spline.evaluate(spline.clampTime(localAnimationTime), runtimeNode[targetPath]);
+                    runtimeNode.dirtyNumber = model._maxDirtyNumber;
```
Comment (lilleyse) [inline Source/Scene/Model.js]: Should the two cases be an if-else in the same function? That way someone would be able to change the `clampAnimation` property dynamically. I don't think there would be much of a performance hit. `this._clampAnimations` could then become `this.clampAnimations` and be a public property.

File: CHANGES.md [hunk 6]
```diff
@@ -7,6 +7,8 @@ Change Log
 * Added ability to support touch event in Imagery Layers Split demo application. [#5948](https://github.com/AnalyticalGraphicsInc/cesium/pull/5948)
```
Comment (lilleyse) [inline CHANGES.md]: This should be simple enough to add to czml too. The only changes would be in `CzmlDataSource`.

File: CHANGES.md [hunk 7]
```diff
@@ -7,6 +7,8 @@ Change Log
...
 * Fixed `Invalid asm.js: Invalid member of stdlib` console error by recompiling crunch.js with latest emscripten toolchain. [#5847](https://github.com/AnalyticalGraphicsInc/cesium/issues/5847)
 * Added CZML support for `polyline.depthFailMaterial`, `label.scaleByDistance`, `distanceDisplayCondition`, and `disableDepthTestDistance`. [#5986](https://github.com/AnalyticalGraphicsInc/cesium/pull/5986)
+* Fixed a bug where models with animations of different lengths would cause an error. [#5694](https://github.com/AnalyticalGraphicsInc/cesium/issues/5694)
+* Added a `clampAnimations` parameter to `Model`. Setting this to false allows different length animations to loop asynchronously.
```
Comment (lilleyse) [inline CHANGES.md]: Also mention that `clampAnimations` is added to `Entity.model`.

File: Source/Scene/Model.js [hunk 8]
```diff
@@ -326,6 +326,7 @@ define([
      * @param {Boolean} [options.allowPicking=true] When <code>true</code>, each glTF mesh and primitive is pickable with {@link Scene#pick}.
      * @param {Boolean} [options.incrementallyLoadTextures=true] Determine if textures may continue to stream in after the model is loaded.
      * @param {Boolean} [options.asynchronous=true] Determines if model WebGL resource creation will be spread out over several frames or block until completion once all glTF files are loaded.
+     * @param {Boolean} [options.clampAnimations=true] Determines if the model's animations should hold a pose over frames where no keyframes are specified.
```
Comment (lilleyse) [inline Source/Scene/Model.js]: Add this parameter to `Model.fromGltf` as well.

**Comment by likangning93** [comment]
@lilleyse updated!

**Comment by lilleyse** [comment]
Looks good. I added a few tweaks in https://github.com/AnalyticalGraphicsInc/cesium/pull/6004/commits/8d458cbf8df9013f9f5375b5aad8ed9229fb97e9.

**Pull request merged**